File: android/src/main/java/com/jtec/idnow/ExpoIDnowModule.kt
package com.jtec.idnow

import de.idnow.sdk.IDnowSDK
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.functions.Coroutine
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException

class ExpoIDnowModule : Module() {
    private val currentActivity
        get() = appContext.activityProvider?.currentActivity ?: throw Exceptions.MissingActivity()

    override fun definition() = ModuleDefinition {
        Name("ExpoIDnow")

        AsyncFunction("startIdent") Coroutine { companyId: String, token: String, options: ExpoIDnowOptions ->
            return@Coroutine startIdent(companyId, token, options)
        }
    }

    private suspend fun startIdent(
        companyId: String, token: String, options: ExpoIDnowOptions
    ): String = withContext(Dispatchers.Main) {
        suspendCancellableCoroutine { continuation ->
            try {
                IDnowSDK.getInstance().initialize(currentActivity, companyId, options.language);
                IDnowSDK.setTransactionToken(token)
                IDnowSDK.setNewBrand(true);
                IDnowSDK.setConnectionType(
                    options.connectionType.toIDnowConnectionType(), currentActivity
                )
                IDnowSDK.setEnvironment(options.environment.toIDnowEnvironment());
                IDnowSDK.setShowErrorSuccessScreen(
                    options.showErrorSuccessScreen, currentActivity
                )
                IDnowSDK.setShowVideoOverviewCheck(
                    options.showVideoOverviewCheck, currentActivity
                )
                if (options.calledFromIDnowApp) {
                    IDnowSDK.calledFromIDnowApp(currentActivity)
                }


                IDnowSDK.getInstance().start(IDnowSDK.getTransactionToken()) { result, _ ->
                    if (!continuation.isCompleted) {
                        continuation.resume(ExpoIDnowResponse(result).toJsonString())
                    }
                }
            } catch (e: Exception) {
                if (!continuation.isCompleted) {
                    continuation.resumeWithException(e)
                }
            }
        }
    }
}

File: ios/ExpoIDnowModule.swift
import ExpoModulesCore
import IDnowSDK

public class ExpoIDnowModule: Module {
	public func definition() -> ModuleDefinition {
		Name("ExpoIDnow")

		AsyncFunction("startIdent") { (
			companyId: String,
			token: String,
			options: ExpoIDnowOptions,
			promise: Promise) in
			if let rootViewController = UIApplication.shared.delegate?.window??.rootViewController {
				let settings = IDnowSettings(companyID: companyId, transactionToken: token)
				settings.userInterfaceLanguage = options.language
				settings.connectionType = options.connectionType.toIDnowConnectionType()
				settings.environment = options.environment.toIDnowEnvironment()
				settings.showErrorSuccessScreen = options.showErrorSuccessScreen
				settings.showVideoOverviewCheck = options.showVideoOverviewCheck
				
				let controller = IDnowController(settings: settings)
				controller.initialize(completionBlock: { _, error, cancelled in
					if error != nil {
						return promise.reject(error ?? Exception(name: "Initialize", description: "IDnow Initialize Error"))
					}
					
					if cancelled {
						return promise.resolve(ExpoIDnowResponse(result: "CANCELED").toJsonString())
					}
					
					controller.startIdentification(from: rootViewController, withCompletionBlock: { _, error, cancelled in
						if error != nil {
							return promise.reject(error ?? Exception(name: "StartIdent", description: "IDnow Start Ident Error"))
						}
						
						if cancelled {
							return promise.resolve(ExpoIDnowResponse(result: "CANCELED").toJsonString())
						}
						
						return promise.resolve(ExpoIDnowResponse(result: "SUCCEED").toJsonString())
					})
				})
			} else {
				return promise.reject(Exception(name: "General", description: "Failed to resolve current Expo Activity"))
			}
		}
	}
}
